File: kAndr/ch_2/fibonacci.c
#include <stdio.h>

// time for fibonacci of 41:  0m1.074s
int fibonacci(int n) {
    if (n <= 1) {
        return 1;
    }
    return fibonacci(n-1) + fibonacci(n-2);
}
// time for fibonacci_dynamic of 41: 0m0.005s
// time for fibonacci_dynamic of 1000: 0m0.003s
long double fibonacci_dynamic(n) {
    long double current, new_current, prev_val; 
    int i;
    current = 1;
    prev_val = 0;
    i = 0;
    while (i < n) {
        new_current = prev_val + current;
        prev_val = current;
        current = new_current;
        i++;
    }
    return current;
}

int main() {
    int n = 1000;
    long double ans;
    ans = fibonacci_dynamic(n);
    printf("Fibonacci of %d is %Lf \n", n, ans);
}

File: kAndr/ch_2/fibonacci.py
#time for fibonacci of 41: 1m14.016s
def fibonacci(n):
    if n <= 1:
        return 1
    return fibonacci(n-1) + fibonacci(n-2)

# print(fibonacci(41))
storage = {1: 1, 0: 1}

#time for fibonacci_optimized of 41: 
def fibonacci_optimized(n):
    if n in storage:
        return storage[n]
    else:
        storage[n] = fibonacci_optimized(n-1) + fibonacci_optimized(n-2)
        return storage[n]
# print(fibonacci_optimized(41))

#time for fib_dynamic of 41: 0m0.034s
#time for fib_dynamic of 1000: 0m0.041s
def fib_dynamic(n):
    current = 1
    prev_val = 0
    i = 0
    while i < n:
        new_current = prev_val + current
        prev_val = current
        current = new_current
        i+=1
    return current
print(fib_dynamic(1000))